<center><img src="https://accredian.com/wp-content/uploads/2023/06/cropped-accredian-ws-logo-e1686627987452.webp" /></center>

---

# <center><b>Accredian Telecom<b></center>

---
# **Table of Contents**
---

1. [**Introduction**](#Section1)<br>
2. [**Problem Statement**](#Section2)<br>
    2.1 [**Project Details**](#Section21)<br>
3. [**Installing & Importing Libraries**](#Section3)<br>
  3.1 [**Installing Libraries**](#Section31)<br>
  3.2 [**Upgrading Libraries**](#Section32)<br>
  3.3 [**Importing Libraries**](#Section33)<br>
4. [**Data Acquisition & Description**](#Section4)<br>
5. [**Data Pre-Profiling**](#Section5)<br>
6. [**Data Pre-Processing**](#Section6)<br>
7. [**Data Post-Profiling**](#Section7)<br>
8. [**Exploratory Data Analysis**](#Section8)<br>
9. [**Summarization**](#Section9)</br>
  9.1 [**Conclusion**](#Section91)</br>
  9.2 [**Actionable Insights**](#Section91)</br>

---

---
<a name = Section1></a>
# **1. Introduction**
---

**Company Introduction**

Our company for this project is AccredianTelecom.

- AccredianTelecom, one of the leading telecom players, understands that customizing offerings is very important for its business to stay competitive.
- Currently, AccredianTelecom is seeking to leverage behavioural data from more than 60% of the 50 million mobile devices active daily in India
- They are doing this to help their clients better understand and interact with their audiences.

**Current Scenario**

- In this consulting assignment, Accredianians are expected to build a dashboard.
- This dashboard will help us to understand a user's demographic characteristics based on their mobile usage, geolocation, and mobile device properties
- Doing so will help millions of developers and brand advertisers around the world pursue data-driven marketing efforts.
- These efforts are relevant to their users and cater to their preferences.

---
<a name = Section2></a>
# **2. Problem Statement**
---


- Fetch the Data onto your Python Notebook and ponder upon the challenges present in the Data.

<a name = Section21></a>
### **2.1 Project Details**

Our Role

- To help the customer the consultants are expected to have depth of clarity in the underlying data.
- How much effort has been put into cleansing and purifying the data will decide how closely have you looked at the data..
- How detailed is the observation stated in the submission report and finally, how well a group presents their consulting journey.
- Please remember that this is an analytics consulting firm. Hence, your efforts in terms of finding user behaviour are going to directly impact the company's offerings.
- Do help the company understand what is the right way forward and suggest actionable insights from marketing and product terms.

---
<a id = Section3></a>
# **3. Installing & Importing Libraries**
---

This section is emphasised on installing and importing the necessary libraries that will be required.

<a name = Section31></a>
### **3.1 Installing Libraries**

In [1]:
!pip install -q mysql-connector-python
# Library to connect to mysql database and fire queries
!pip install -q ydata-profiling
# Library to generate basic statistics about data

# To install more libraries insert your code here..

print("Completed!!")

Completed!!


In [2]:
!pip install geopy
print("Completed!")

Defaulting to user installation because normal site-packages is not writeable
Completed!


<a name = Section32></a>
### **3.2 Upgrading Libraries**

- **After upgrading** the libraries, you need to **restart the runtime** to make the libraries in sync.

- Ensure not to execute the cell under Installing Libraries and Upgrading Libraries again after restarting the runtime.

In [3]:
!pip install -q --upgrade ydata-profiling
# Upgrading pandas profiling to the latest version

print("Completed!!")

Completed!!


<a name = Section33></a>
### **3.3 Importing Libraries**
- Installing the necessary libraries, packages - 
    - Pandas, 
    - Numpy, 
    - Matplotlib, 
    - Seaborn, 
    - scipy, 
    - mysql_connector, etc.

In [4]:
import mysql.connector as sql
# Importing mysql connector bridge

import pandas as pd
# Importing package pandas (For Panel Data Analysis)

import numpy as np
# Importing package numpys (For Numerical Python)

import matplotlib.pyplot as plt
# Importing pyplot interface to use matplotlib

import seaborn as sns
 # Importing seaborn library for interactive visualization

%matplotlib inline

import scipy as sp
# Importing library for scientific calculations

print("Completed!")

Completed!


In [5]:
from geopy.geocoders import Nominatim
print("Completed!")

Completed!


In [6]:
from ydata_profiling import ProfileReport
# Import Pandas Profiling (To generate Univariate Analysis)
print("Completed!")

Completed!


---
<a name = Section4></a>
# **4. Data Acquisition & Description**
---
- In this assignment, we are going to study the demographics of a user (gender and age) based on their app download and usage behaviours.
- The Data is collected from mobile apps that use Inside Telecom services.
- Full recognition and consent from an individual user of those apps have been obtained, and appropriate anonymization has been performed to protect privacy.
- Due to confidentiality, we won't provide details on how the gender and age data were obtained.
- Please treat them as accurate ground truth for prediction.
- The data schema can be represented in the following table:

    - gender_age_train - Devices and their respective user gender, age and age_group
    - phone_brand_device_model - device ids, brand, and models phone_brand: note that few brands are in Chinese

|Id|Brand Name|English Mapping|
|:--|:--|:--|
|01|华为|Huawei|
|02|小米|Xiaomi|
|03|三星|Samsung|
|04|vivo|vivo|
|05|OPPO|OPPO|
|06|魅族|Meizu|
|07|酷派|Coolpad|
|08|乐视|LeEco|
|09|联想|Lenovo|
|10|HTC|HTC|

- Events_data:
    
    - When a user uses mobile on an Accredian Telecom network, the event gets logged in this data.
    - Each event has an event id, location (lat/long), and the event corresponds to the frequency of mobile usage. timestamp: when the user is using the mobile.

In [7]:
connectionObj = sql.connect(host="148.72.246.142", database = "Capstone1", user = "student", password = "student")
# To Create Connection
print("DB Connection successful!")

DB Connection successful!


In [8]:
# csr1_gen = connectionObj.cursor()
gender_age_train_df=pd.read_sql_query("SELECT * FROM gender_age_train ",connectionObj)
print("Gender Age Train Data Set Loading completed!")
print("_____________________________________________________________________")
print("Dimensions of Gender Age Train Data is ",gender_age_train_df.shape)

# csr2_phon = connectionObj.cursor()
phone_brand_device_df=pd.read_sql_query("SELECT * FROM phone_brand_device_model",connectionObj)
print("Phone Brand Device Model Data Set loading is completed!!")
print("___________________________________________________________________________")
print("Dimensions Phone Brand and Device Model Data is ",phone_brand_device_df.shape)


C:\Users\Lenovo\AppData\Local\Temp\ipykernel_1468\1818854316.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  gender_age_train_df=pd.read_sql_query("SELECT * FROM gender_age_train ",connectionObj)


Gender Age Train Data Set Loading completed!
_____________________________________________________________________
Dimensions of Gender Age Train Data is  (74645, 4)


C:\Users\Lenovo\AppData\Local\Temp\ipykernel_1468\1818854316.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  phone_brand_device_df=pd.read_sql_query("SELECT * FROM phone_brand_device_model",connectionObj)


Phone Brand Device Model Data Set loading is completed!!
___________________________________________________________________________
Dimensions Phone Brand and Device Model Data is  (87726, 3)


In [9]:
gender_age_train_df.head()

device_id gender  age   group
0 -8076087639492063270      M   35  M32-38
1 -2897161552818060146      M   35  M32-38
2 -8260683887967679142      M   35  M32-38
3 -4938849341048082022      M   30  M29-31
4   245133531816851882      M   30  M29-31

In [10]:
phone_brand_device_df.head()

device_id phone_brand device_model
0  1877775838486905855        vivo          Y13
1 -3766087376657242966          小米         V183
2 -6238937574958215831        OPPO          R7s
3  8973197758510677470          三星        A368t
4 -2015528097870762664          小米      红米Note2

In [11]:
# from google.colab import drive
# drive.mount('/content/drive')

Data Set Path for Event file:

/content/drive/MyDrive/Colab Notebooks/Data Set/events_data.csv

In [14]:
event_df = pd.read_csv("../dataset/events_data.csv")
print("Events Data Set loading is completed!!!")
print("_____________________________________________________________")
print("Dimensions of Events Data is ", event_df.shape)

Events Data Set loading is completed!!!
_____________________________________________________________
Dimensions of Events Data is  (3252950, 7)


In [15]:
event_df.head()

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968 -3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   

       city       state  
0     Delhi       Delhi  
1  Calcutta  WestBengal  
2     Delhi       Delhi  
3   Chennai   TamilNadu  
4    Bokaro   Jharkhand

In [16]:
print("Gender age train ->",gender_age_train_df.shape)
print("Phone Brand Device Model ->",phone_brand_device_df.shape)
print("Event ->",event_df.shape)


Gender age train -> (74645, 4)
Phone Brand Device Model -> (87726, 3)
Event -> (3252950, 7)


<a name = Section41></a>
### **4.1 Data Description**


#### **gender_age_train**

In [17]:
gender_age_train_df.describe()

device_id           age
count  7.464500e+04  74645.000000
mean  -7.491354e+14     31.410342
std    5.327150e+18      9.868735
min   -9.223067e+18      1.000000
25%   -4.617367e+18     25.000000
50%   -1.841362e+16     29.000000
75%    4.636656e+18     36.000000
max    9.222849e+18     96.000000

#### **phone_brand_device_model**

In [18]:
phone_brand_device_df.describe()

device_id
count  8.772600e+04
mean  -5.212557e+15
std    5.330157e+18
min   -9.223322e+18
25%   -4.623331e+18
50%   -3.109802e+16
75%    4.635194e+18
max    9.222849e+18

#### **Events_data**

In [19]:
event_df.describe()

event_id     device_id     longitude      latitude
count  3.252950e+06  3.252497e+06  3.252527e+06  3.252527e+06
mean   1.626476e+06  1.012200e+17  7.815868e+01  2.168851e+01
std    9.390459e+05  5.316758e+18  4.235570e+00  5.789111e+00
min    1.000000e+00 -9.222957e+18  1.256700e+01  8.190110e+00
25%    8.132382e+05 -4.540611e+18  7.583543e+01  1.780171e+01
50%    1.626476e+06  1.726820e+17  7.726814e+01  2.216454e+01
75%    2.439713e+06  4.861813e+18  8.031916e+01  2.868278e+01
max    3.252950e+06  9.222849e+18  9.545946e+01  4.187190e+01

<a name = Section42></a>
### **4.2 Data Information**

#### **gender_age_train**

In [20]:
gender_age_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  74645 non-null  int64 
 1   gender     74645 non-null  object
 2   age        74645 non-null  int64 
 3   group      74645 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.3+ MB


#### **phone_brand_device_model**

In [21]:
phone_brand_device_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   device_id     87726 non-null  int64 
 1   phone_brand   87726 non-null  object
 2   device_model  87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB


#### **Events_data**

In [22]:
event_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype  
---  ------     -----  
 0   event_id   int64  
 1   device_id  float64
 2   timestamp  object 
 3   longitude  float64
 4   latitude   float64
 5   city       object 
 6   state      object 
dtypes: float64(3), int64(1), object(3)
memory usage: 173.7+ MB


---
<a name = Section5></a>
# **5. Data Pre-Profiling**
---

In [23]:
# profile = ProfileReport(df = event_df, minimal = False, explorative=True, dark_mode=True, title = "Events Pre Profiling Report")
# profile.to_file(output_file='Event DF Pre profiling report.html')
# print('Profiling Completed!!!')

In [24]:
# profile = ProfileReport(df = gender_age_train_df, minimal = False, explorative=True, dark_mode=True, title = "Gender Age Pre Profiling Report")
# profile.to_file(output_file='Gender age Pre profiling report.html')
# print('Profiling Completed!!!')

In [25]:
# profile = ProfileReport(df = phone_brand_device_df, minimal = False, explorative=True, dark_mode=True, title = "Phone Brand Device Pre-profiling Report")
# profile.to_file(output_file='Phone Brand Device Pre profiling report.html')
# print('Profiling Completed!!!')

In [26]:
event_df[event_df['state']=='Maharashtra']['device_id'].value_counts()

 4.428421e+17    6718
 2.601607e+18    3425
-5.829328e+18    2861
-7.364841e+18    2625
 7.018600e+18    1915
                 ... 
-3.653909e+18       1
 5.113546e+18       1
 7.885046e+18       1
-6.058409e+18       1
-1.420493e+18       1
Name: device_id, Length: 9010, dtype: int64

In [27]:
event_df['state'].value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
Meghalaya                     1515
AndamanandNicobarIslands      1515
HimachalPradesh               1409
Manipur             

In [28]:
phone_brand_device_df['device_id']

0        1877775838486905855
1       -3766087376657242966
2       -6238937574958215831
3        8973197758510677470
4       -2015528097870762664
                ...         
87721   -4961458925928574032
87722   -8819817317449262375
87723   -3358291377416935026
87724    3282788959750981633
87725    2491639413207286271
Name: device_id, Length: 87726, dtype: int64

### Missing Values analysis

In [29]:
event_df.isna().sum()

event_id       0
device_id    453
timestamp      0
longitude    423
latitude     423
city           0
state        377
dtype: int64

Missing Values are as follows:
- device_id - 453
- longitude - 423
- latitude - 423
- state - 377

In [30]:
phone_brand_device_df.isna().sum()

device_id       0
phone_brand     0
device_model    0
dtype: int64

In [31]:
gender_age_train_df.isna().sum()

device_id    0
gender       0
age          0
group        0
dtype: int64

### Duplicated Entries/Values

In [32]:
event_df.duplicated().value_counts()

False    3252950
dtype: int64

In [33]:
phone_brand_device_df.duplicated().value_counts()

False    87726
dtype: int64

In [34]:
gender_age_train_df.duplicated().value_counts()

False    74645
dtype: int64

### Handling Outliers

### Improper Data Type

### **Observations**

---
<a name = Section6></a>
# **6. Data Cleaning**
---

Missing Values

In [35]:
event_df.head()

event_id     device_id            timestamp  longitude   latitude  \
0   2765368  2.973348e+18  2016-05-07 22:52:05  77.225676  28.730140   
1   2955066  4.734221e+18  2016-05-01 20:44:16  88.388361  22.660325   
2    605968 -3.264500e+18  2016-05-02 14:23:04  77.256809  28.757906   
3    448114  5.731369e+18  2016-05-03 13:21:16  80.343613  13.153332   
4    665740  3.388880e+17  2016-05-06 03:51:05  85.997745  23.842609   

       city       state  
0     Delhi       Delhi  
1  Calcutta  WestBengal  
2     Delhi       Delhi  
3   Chennai   TamilNadu  
4    Bokaro   Jharkhand

In [36]:
event_df[event_df['device_id'].isna()]

event_id  device_id            timestamp  longitude   latitude  \
2055      1936567        NaN  2016-05-02 15:27:54  75.882956  22.814519   
26367     2659282        NaN  2016-05-04 14:06:24  75.888487  26.948689   
42088      442198        NaN  2016-05-06 09:27:37  75.846007  26.960796   
43061     1903648        NaN  2016-05-02 15:41:54  75.923332  22.777781   
44823     3227383        NaN  2016-05-05 08:37:09  75.992551  31.561747   
...           ...        ...                  ...        ...        ...   
3212337   2191286        NaN  2016-05-04 18:16:06  83.357991  17.805195   
3215120    572054        NaN  2016-05-07 08:48:22  83.371738  17.752819   
3230115    332451        NaN  2016-05-07 09:01:45  75.992551  31.561747   
3235494   2534073        NaN  2016-05-06 21:07:01  83.371738  17.752819   
3250939   1622538        NaN  2016-05-06 21:26:10  75.888487  26.948689   

                  city          state  
2055            Indore  MadhyaPradesh  
26367           Jaipur      Rajasthan  
42088           Jaipur      Rajasthan  
43061           Indore  MadhyaPradesh  
44823       Hoshiarpur         Punjab  
...                ...            ...  
3212337  Visakhapatnam  AndhraPradesh  
3215120  Visakhapatnam  AndhraPradesh  
3230115     Hoshiarpur         Punjab  
3235494  Visakhapatnam  AndhraPradesh  
3250939         Jaipur      Rajasthan  

[453 rows x 7 columns]

In [37]:
event_df['device_id'].mode()

0    5.536513e+18
Name: device_id, dtype: float64

In [40]:
event_df.state.value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
Meghalaya                     1515
AndamanandNicobarIslands      1515
HimachalPradesh               1409
Manipur             

In [41]:
event_df[event_df.state =='Maharashtra']

event_id     device_id            timestamp  longitude   latitude  \
9          720049  4.928556e+18  2016-05-06 19:00:52  72.844795  18.974929   
11          13998  3.978026e+18  2016-05-06 13:12:42  73.875196  18.615791   
20        1378252 -4.384474e+18  2016-05-05 04:20:02  73.914179  18.619081   
21         185325 -7.369690e+18  2016-05-02 20:07:07  72.874552  19.014537   
22        1861432  3.965811e+18  2016-05-02 20:49:13  73.938826  18.582846   
...           ...           ...                  ...        ...        ...   
3252938     73002 -5.758560e+18  2016-05-03 06:30:02  73.858068  18.574054   
3252941   3046303  5.212597e+18  2016-05-04 07:50:48  73.897710  18.565680   
3252942   3244031  2.320543e+18  2016-05-07 18:36:01  73.886764  18.585058   
3252945   2687452 -1.937028e+18  2016-05-07 23:33:14  73.891597  18.544124   
3252946   1051580  3.345851e+18  2016-05-03 05:13:30  72.837258  19.018432   

           city        state  
9        Mumbai  Maharashtra  
11         Pune  Maharashtra  
20         Pune  Maharashtra  
21       Mumbai  Maharashtra  
22         Pune  Maharashtra  
...         ...          ...  
3252938    Pune  Maharashtra  
3252941    Pune  Maharashtra  
3252942    Pune  Maharashtra  
3252945    Pune  Maharashtra  
3252946  Mumbai  Maharashtra  

[677105 rows x 7 columns]

### Changing the device_id dtype to object.

In [42]:
event_df['device_id'] = event_df['device_id'].astype(object)

In [50]:
event_df['device_id'].mode()[0]

5.536513450525271e+18

In [45]:
event_df['device_id'].fillna(event_df['device_id'].mode()[0], inplace=True)

In [46]:
event_df['device_id']

0          2.973348e+18
1          4.734221e+18
2         -3.264500e+18
3          5.731369e+18
4          3.388880e+17
               ...     
3252945   -1.937028e+18
3252946    3.345851e+18
3252947   -6.406040e+18
3252948   -2.920741e+18
3252949    3.212750e+18
Name: device_id, Length: 3252950, dtype: float64

In [47]:
geolocator = Nominatim(user_agent="Acc Telecom")

In [48]:
def geocode(city):
    location = geolocator.geocode(city)
    return location.latitude, location.longitude

In [49]:
for index, row in event_df.iterrows():
    if pd.isnull(row['latitude']) or pd.isnull(row['longitude']):
        latitude, longitude = geocode(row['city'])
        event_df.at[index, 'latitude'] = latitude
        event_df.at[index, 'longitude'] = longitude

KeyboardInterrupt: 

---
<a name = Section7></a>
# **7. Data Post-Profiling**
---

---
<a name = Section8></a>
# **8. Exploratory Data Analysis**
---

---
<a name = Section9></a>
# **9. Summarization**
---

<a name = Section91></a>
### **9.1 Conclusion**

<a name = Section92></a>
### **9.2 Actionable Insights**